dask/array/slicing.py

```diff
@@ -1335,3 +1338,847 @@ def cached_cumsum(seq, initial_zero=False):
+            value_indices.insert(0, Ellipsis)
+
+        # Create the part of the full assignment value that is to be
+        # assigned to elements of this block and make sure that it has
+        # just one chunk (so we can represent it with a single key in
+        # the argument list of setitem).
+        v = value[tuple(value_indices)]
+        v = concatenate_array_chunks(v)
+        v_key = next(flatten(v.__dask_keys__()))
+
+        # Insert into the output dask dictionary the dask of the part
+        # of assignment value for this block (not minding when we
+        # overwrite any existing keys as the values will be the same).
+        dsk = merge(dict(v.dask), dsk)
+
+        # Define the assignment function for this block.
+        dsk[out_key] = (setitem, in_key, v_key, block_indices)
+
+    block_index_from_1d_index.cache_clear()
+    block_index_shape_from_1d_bool_index.cache_clear()
+    n_preceeding_from_1d_bool_index.cache_clear()
+    value_indices_from_1d_int_index.cache_clear()
+
+    return dsk
+
+
+def setitem(x, v, indices):
+    """Chunk function of `setitem_array`.
+
+    Assign v to indices of x.
+
+    Parameters
+    ----------
+    x : numpy array
+        The array to be assigned to.
+    v : numpy array
+        The values which will be assigned.
+    indices : list of `slice`, `int`, or numpy array
+        The indices describing the elements of x to be assigned from
+        v. One index per axis.
+
+        Note that an individual index can not be a `list`, use a 1-d
+        numpy array instead.
+
+        If a 1-d numpy array index contains the non-valid value of the
+        size of the corresponding dimension of x, then those index
+        elements will be removed prior to the assignment (see
+        `block_index_from_1d_index` function).
+
+    Returns
+    -------
+    numpy array
+        A new independent array with assigned elements, unless v is
+        empty (i.e. has zero size) in which case then the input array
+        is returned and the indices are ignored.
+
+    Examples
+    --------
+    >>> x = np.arange(8).reshape(2, 4)
+    >>> setitem(x, -99, [np.array([False, True])])
+    array([[  0,   1,   2,   3],
+           [-99, -99, -99, -99]])
+    >>> x
+    array([[ 0,  1,  2,  3],
+           [ 4,  5,  6,  7]])
+
+    >>> setitem(x, [-88, -99], [slice(None), np.array([1, 3])])
+    array([[  0, -88,   2, -99],
+           [  4, -88,   6, -99]])
+
+    >>> setitem(x, -x, [slice(None)])
+    array([[  0,  -1,  -2,  -3],
+           [ -4,  -5,  -6,  -7]])
+    >>> x
+    array([[ 0,  1,  2,  3],
+           [ 4,  5,  6,  7]])
+
+    >>> setitem(x, [-88, -99], [slice(None), np.array([4, 4, 3, 4, 1, 4])])
+    array([[  0, -99,   2, -88],
+           [  4, -99,   6, -88]])
+
+    >>> value = np.where(x < 0)[0]
+    >>> value.size
+    0
+    >>> y = setitem(x, value, [Ellipsis])
+    >>> y is x
+    True
+
+    """
+    if not v.size:
```

Comment:
The `v.size` seems to be breaking the doctests, because `v=-99`. 

Is there supposed to be a `v = np.asarray(v)` before the check or are the doctests outdated and this function should only handle numpy arrays?

Reply:
Good spot. The examples are out of date, as `v` is now expected to be numpy array. I have updated docstring.

Reply:
It'd be good to include an `asarray` here or look for the attr `size` so that you can raise a sensible exception if `v` does not have `size`

Reply:
Hi @jsignell, I'm fine with putting in a `v = np.asanyarray(v)` into the  first line of `setitem`, but as `v` is coming from the resolution of a dask key, is it not already guaranteed to a numpy array?

In the following:
```python
('setitem-d7e36a', 0, 1): (setitem, ('array-836255', 0, 1), ('getitem-5ad1e3', 0, 0), [slice(2, 4, 1), slice(0, 4, 1)])
```
`('getitem-5ad1e3', 0, 0)` "is" `v`.

 Apologies if I've misunderstood!

Reply:
No you are right. I think I read too quickly.